src/main/java/com/exactpro/th2/crawler/Crawler.java

```diff
@@ -52,205 +52,238 @@
 
-                    if (lastProcessedEvent != null) {
+                lastProcessedEvent = state == null ? null : state.getLastProcessedEvent();
+
+                EventID startId = null;
+
+                if (lastProcessedEvent != null) {
+                    if (!fetchIntervalReport.processFromStart) {
                         startId = EventUtils.toEventID(lastProcessedEvent.getId());
                     }
+                }
 
-                    report = sendEvents(crawlerId, info, batchSize, startId);
+                sendingReport = sendEvents(new EventsInfo(interval, info, startId,
+                        interval.getStartTime(), interval.getEndTime()));
 
-                } else if (MESSAGES.equals(interval.getCrawlerType())) {
-                    report = sendMessages(crawlerId, info, batchSize);
-                } else {
-                    throw new ConfigurationException("Type must be either EVENTS or MESSAGES");
-                }
+            } else if (MESSAGES.equals(interval.getCrawlerType())) {
+                Map<String, RecoveryState.InnerMessage> lastProcessedMessages;
+                RecoveryState state = RecoveryState.getStateFromJson(interval.getRecoveryState());
 
-                restartPod = report.action == CrawlerAction.STOP;
+                lastProcessedMessages = state == null ? null : state.getLastProcessedMessages();
 
-                if (report.action == CrawlerAction.NONE) {
-                    interval = intervalsWorker.setIntervalProcessed(interval, true);
+                Map<String, MessageID> startIds = null;
 
-                    RecoveryState previousState = interval.getRecoveryState();
+                MessageID.Builder builder = MessageID.newBuilder();
 
-                    if (EVENTS.equals(interval.getCrawlerType())) {
-                        RecoveryState state = new RecoveryState(
-                                null,
-                                previousState.getLastProcessedMessages(),
-                                numberOfEvents,
-                                previousState.getLastNumberOfMessages());
+                if (lastProcessedMessages != null) {
+                    if (!fetchIntervalReport.processFromStart) {
+                        List<MessageID> ids = lastProcessedMessages.values().stream()
+                                .map(innerMessage -> {
+                                    com.exactpro.th2.common.grpc.Direction direction;
 
-                        interval = intervalsWorker.updateRecoveryState(interval, state);
-                    }
+                                    if (innerMessage.getDirection() == Direction.FIRST)
+                                        direction = com.exactpro.th2.common.grpc.Direction.FIRST;
+                                    else
+                                        direction = com.exactpro.th2.common.grpc.Direction.SECOND;
+
+                                    return builder
```

Comment:
What about subsequences? Shouldn't we store them as well?

Reply:
As far as I remember, during one of the calls Nikita told me that depending on subsequences was not a good idea. Crawler should start from the first subsequence after recovering.